xarray/core/duck_array_ops.py

```diff
@@ -279,14 +279,49 @@ def f(values, axis=None, skipna=None, **kwargs):
 _mean = _create_nan_agg_method('mean')
 
 
+def datetime_to_numeric(array, offset=None, datetime_unit=None, dtype=float):
+    """Convert an array containing datetime-like data to an array of floats.
+
+    Parameters
+    ----------
+    da : array
+        Input data
+    offset: Scalar with the same type of array or None
+        If None, subtract minimum values to reduce round off error
+    datetime_unit: None or any of {'Y', 'M', 'W', 'D', 'h', 'm', 's', 'ms',
+        'us', 'ns', 'ps', 'fs', 'as'}
+    dtype: target dtype
+
+    Returns
+    -------
+    array
+    """
+    if offset is None:
+        offset = array.min()
+    array = array - offset
+
+    if datetime_unit:
+        array = array / np.timedelta64(1, datetime_unit)
+    # convert np.NaT to np.nan
+    if hasattr(array, 'dtype'):
+        if array.dtype.kind in 'mM':
+            return np.where(isnull(array), np.nan, array.astype(dtype))
+        else:
+            print(type(array), array.dtype)
+            print(array)
+            return array.astype(dtype)
```

Comment:
Some tests pass a `np.ndarray` with object dtype, an entry of which is `datetime.datetime`, such as
```python
<class 'numpy.ndarray'> object
[datetime.timedelta(0) datetime.timedelta(days=1)
 datetime.timedelta(days=2)]
```
Is it an expected behavior?

Reply:
I don't think `astype()` does any conversion for object arrays.

Reply:
@fujiisoup this function might be helpful for casting the object arrays: https://github.com/pydata/xarray/blob/d4c46829b283ab7e7b7db8b86dae77861ce68f3c/xarray/core/variable.py#L132-L136